packages/world/src/Vec3.sol

```diff
@@ -0,0 +1,258 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.24;
+
+import { Position } from "./codegen/tables/Position.sol";
+import { ReversePosition } from "./codegen/tables/ReversePosition.sol";
+import { Direction } from "./codegen/common.sol";
+import { CHUNK_SIZE, FORCE_FIELD_SHARD_DIM, LOCAL_ENERGY_POOL_SHARD_DIM } from "./Constants.sol";
+
+// Vec3 stores 3 packed int32 values (x, y, z)
+type Vec3 is uint96;
+
+function vec3(int32 _x, int32 _y, int32 _z) pure returns (Vec3) {
+  // Pack 3 int32 values into a single uint96
+  return Vec3.wrap((uint96(uint32(_x)) << 64) | (uint96(uint32(_y)) << 32) | uint96(uint32(_z)));
+}
+
+function eq(Vec3 a, Vec3 b) pure returns (bool) {
+  return Vec3.unwrap(a) == Vec3.unwrap(b);
+}
+
+function neq(Vec3 a, Vec3 b) pure returns (bool) {
+  return Vec3.unwrap(a) != Vec3.unwrap(b);
+}
+
+function add(Vec3 a, Vec3 b) pure returns (Vec3) {
+  return vec3(a.x() + b.x(), a.y() + b.y(), a.z() + b.z());
+}
+
+function sub(Vec3 a, Vec3 b) pure returns (Vec3) {
+  return vec3(a.x() - b.x(), a.y() - b.y(), a.z() - b.z());
+}
+
+function getDirectionVector(Direction direction) pure returns (Vec3) {
+  if (direction == Direction.PositiveX) return vec3(1, 0, 0);
+  if (direction == Direction.NegativeX) return vec3(-1, 0, 0);
+  if (direction == Direction.PositiveY) return vec3(0, 1, 0);
```

Comment:
not sure how often this is used but would it be worth pre-computing these instead of encoding on the fly?

Reply:
yeah we can definitely do that, I was just lazy, but not used that much